src/thread_task.rs

```diff
@@ -30,265 +28,56 @@ use crate::internal::{
-        while t.is_some() {
-            'next: {
-                // entropy coding precedes other steps
-                if self.index(t).type_0 == TaskType::TileEntropy {
-                    if self.index(first).type_0 > TaskType::TileEntropy {
-                        break 'next;
-                    }
-                    // both are entropy
-                    if self.index(first).sby > self.index(t).sby {
-                        break 'next;
-                    }
-                    if self.index(first).sby < self.index(t).sby {
-                        self.insert_tasks_between(c, first, last, prev_t, t, cond_signal);
-                        return;
-                    }
-                    // same sby
-                } else {
-                    if self.index(first).type_0 == TaskType::TileEntropy {
-                        self.insert_tasks_between(c, first, last, prev_t, t, cond_signal);
-                        return;
-                    }
-                    if self.index(first).sby > self.index(t).sby {
-                        break 'next;
-                    }
-                    if self.index(first).sby < self.index(t).sby {
-                        self.insert_tasks_between(c, first, last, prev_t, t, cond_signal);
-                        return;
-                    }
-                    // same sby
-                    if self.index(first).type_0 > self.index(t).type_0 {
-                        break 'next;
-                    }
-                    if (self.index(first).type_0) < self.index(t).type_0 {
-                        self.insert_tasks_between(c, first, last, prev_t, t, cond_signal);
-                        return;
-                    }
-                    // same task type
-                }
-
-                // sort by tile-id
-                assert!(
-                    self.index(first).type_0 == TaskType::TileReconstruction
-                        || self.index(first).type_0 == TaskType::TileEntropy
-                );
-                assert!(self.index(first).type_0 == self.index(t).type_0);
-                assert!(self.index(t).sby == self.index(first).sby);
-                let t_tile_idx = self.index(first).tile_idx;
-                let p_tile_idx = self.index(t).tile_idx;
-                assert!(t_tile_idx != p_tile_idx);
-                if !(t_tile_idx > p_tile_idx) {
-                    self.insert_tasks_between(c, first, last, prev_t, t, cond_signal);
-                    return;
-                }
-            }
-            // next:
-            prev_t = t;
-            t = self.index(t).next();
+    fn push(&self, c: &Rav1dContext, task: Rav1dTask, cond_signal: c_int) {
+        let frame_idx = task.frame_idx;
+        {
+            let mut tasks = self.tasks.try_write().unwrap();
+            let idx = match tasks.binary_search(&task) {
+                Ok(_) => panic!("Tried to add an identical task!"),
+                Err(idx) => idx,
+            };
```

Comment:
The code didn't use to do this, right?  Maybe just make it a `debug_assertions`-only check?